wave_utils/include/wave/utils/config.hpp

```diff
@@ -131,93 +128,125 @@ class ConfigParser {
+    /** Convert YAML node to Eigen Matrix
      *
-     * @todo refactor return codes into an enum which can be documented
+     * @throws YAML::InvalidNode if nested keys not found
+     * @returns true if conversion successful
      */
-    int checkVector(std::string key, enum ConfigDataType type, bool optional);
-
-    /** @return see `checkKey` */
-    int checkMatrix(std::string key, bool optional);
-
-    /** Load yaml param primitive, array, vector or matrix.
-     * @return
-     * - `0`: On success
-     * - `-1`: Config file is not loaded
-     * - `-2`: `key` not found in yaml file, parameter is not optional
-     * - `-3`: `key` not found in yaml file, parameter is optional
-     * - `-4`: Invalid vector (wrong size)
-     * - `-5`: Invalid matrix (missing yaml key 'rows', 'cols' or 'data' for
-     * matrix)
-     * - `-6`: Invalid param type
+    static bool decode(const Node &node,
+                       Eigen::Matrix<Scalar, Rows, Cols> &out);
+};
+
+/** Custom conversion functions for YAML -> Eigen vector
+ *
+ * For example:
+ * ```yaml
+ * some_vector: [1.1, 2.2, 3.3]
+ * ```
+ *
+ * Conversions for the following types are compiled as part of wave_utils:
+ * - Eigen::Vector2d
+ * - Eigen::Vector3d
+ * - Eigen::Vector4d
+ * - Eigen::VectorXd
+ */
+template <typename Scalar, int Rows>
+struct convert<Eigen::Matrix<Scalar, Rows, 1>> {
+    /** Convert YAML node to Eigen Matrix
+     * @returns true if conversion successful
      */
-    int loadPrimitive(ConfigParam param);
-    int loadArray(ConfigParam param);
-    int loadVector(ConfigParam param);
-    int loadMatrix(ConfigParam param);
-
-    /** Load yaml file at `config_file`.
-     * @return
-     * - `0`: On success
-     * - `1`: File not found
-     * - `-1`: Config file is not loaded
-     * - `-2`: `key` not found in yaml file
-     * - `-4`: Invalid vector (wrong size)
-     * - `-5`: Invalid matrix (missing yaml key 'rows', 'cols' or 'data' for
-     * matrix)
-     * - `-6`: Invalid param type
+    static bool decode(const Node &node, Eigen::Matrix<Scalar, Rows, 1> &out);
+};
+
+// Explicit instantiations: include these in the compiled wave_utils library
+// Since the function definition is in a .cpp file, other types will not work
+template struct convert<wave::Mat2>;
+template struct convert<wave::Mat3>;
+template struct convert<wave::Mat4>;
+template struct convert<wave::MatX>;
+template struct convert<wave::Vec2>;
+template struct convert<wave::Vec3>;
+template struct convert<wave::Vec4>;
+template struct convert<wave::VecX>;
+
+
+/** Custom conversion for YAML -> cv::Mat
+ *
+ * @note We require the yaml node to have three nested keys in order to parse a
+ * matrix. They are `rows`, `cols` and `data`.
+ *
+ * For example:
+ * ```yaml
+ * some_matrix:
+ *   rows: 3
+ *   cols: 3
+ *   data: [1.1, 2.2, 3.3,
+ *          4.4, 5.5, 6.6,
+ *          7.7, 8.8, 9.9]
+ * ```
+ *
+
+ */
+template <>
+struct convert<cv::Mat> {
```

Comment:
Was this not supposed to be in the vision module?

Reply:
Separate PR #237